### PR TITLE
Build application bundle for macOS

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Application bundles are the user-friendly way of packaging apps on macOS. This PR is working towards including such builds for OpenLoco.

To do:
- [x] Use the OpenLoco icon in the bundle.
- [x] Ship a plist for identification.
- [x] Include changelog and contributor files.
- [x] Automatically upload to tagged releases on GitHub?

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
I think it would be better to build the icns-file at compile-time: https://elliotekj.com/2014/05/27/how-to-create-high-resolution-icns-files/

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,10 @@
 artefacts/
 bin.*/
+boost/
+distribution/macos/AppIcon.iconset/icon_*.png
+distribution/macos/AppIcon.icns
 lib/
+vcpkg/
```
Comment (marijnvdwerf) [inline .gitignore]: just have this and boost be a subfolder of lib

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,10 @@
 artefacts/
 bin.*/
+boost/
+distribution/macos/AppIcon.iconset/icon_*.png
+distribution/macos/AppIcon.icns
```
Comment (marijnvdwerf) [inline .gitignore]: Can't this be generated in the build dir?

File: CMakeLists.txt [hunk 3]
```diff
@@ -230,6 +231,61 @@ endif()
...
+    endif()
+
+    # Create as a bundle
+    set_target_properties(${PROJECT} PROPERTIES
+        MACOSX_BUNDLE ON
+        MACOSX_BUNDLE_INFO_PLIST ${CMAKE_CURRENT_SOURCE_DIR}/distribution/macos/Info.plist.in
+        RESOURCE "${BUNDLE_RESOURCES}")
+
+    # Command to create the directory structure in resources
+    add_custom_command(OUTPUT
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: You can do this using 

```
set_property(
        SOURCE ${OLOCO_RESOURCES}
        PROPERTY MACOSX_PACKAGE_LOCATION "Resources/language"
    )
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -230,6 +231,61 @@ endif()
...
     target_link_libraries(${PROJECT} "-framework Cocoa")
 
+    set(BUNDLE_RESOURCE "${PROJECT}.app/Contents/Resources")
+    set(SOURCE_DATA_DIR "${CMAKE_CURRENT_SOURCE_DIR}/data")
+    set(SOURCE_ICON_DIR "${CMAKE_CURRENT_SOURCE_DIR}/resources/logo")
+    set(ICON_TARGET "${CMAKE_CURRENT_SOURCE_DIR}/distribution/macos/AppIcon.iconset")
+    set(ICON_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/distribution/macos/AppIcon.icns")
+
+    add_custom_command(OUTPUT ${ICON_OUTPUT}
+        COMMAND sips -z 16 16   icon_x16.png  --out ${ICON_TARGET}/icon_16x16.png
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Why are we resampling existing images?

Comment (AaronVanGeffen) [reply CMakeLists.txt]: Might as well copy them over, I guess.

File: .gitignore (hunk 2 above)
Comment (AaronVanGeffen) [reply .gitignore]: Sure, the icns file might as well be. The `AppIcon.iconset` dir contains a Contents.json file, though. I suppose we could copy that over too, though.

**Comment by AaronVanGeffen** [comment]
Right, this PR is in dire need of squashing, but I think that's everything.

@marijnvdwerf Could you verify the build works on your end?

File: CMakeLists.txt [hunk 5]
```diff
@@ -230,6 +231,58 @@ endif()
 if (APPLE)
     target_link_libraries(${PROJECT} "-framework Cocoa")
 
+    file(GLOB OLOCO_RESOURCES "${CMAKE_CURRENT_SOURCE_DIR}/data/language/*")
+
+    set(BUNDLE_RESOURCE "${PROJECT}.app/Contents/Resources")
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Not used

File: CMakeLists.txt [hunk 6]
```diff
@@ -230,6 +231,58 @@ endif()
 if (APPLE)
     target_link_libraries(${PROJECT} "-framework Cocoa")
 
+    file(GLOB OLOCO_RESOURCES "${CMAKE_CURRENT_SOURCE_DIR}/data/language/*")
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Rename to languages, move down to other language stuff

File: CMakeLists.txt [hunk 7]
```diff
@@ -230,6 +231,58 @@ endif()
...
+
+    set(BUNDLE_RESOURCE "${PROJECT}.app/Contents/Resources")
+    set(SOURCE_DATA_DIR "${CMAKE_CURRENT_SOURCE_DIR}/data")
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Unused

File: CMakeLists.txt [hunk 8]
```diff
@@ -230,6 +231,58 @@ endif()
...
+        COMMAND cp icon_x128.png  ${ICON_TARGET}/icon_128x128.png
+        COMMAND cp icon_x256.png  ${ICON_TARGET}/icon_128x128@2x.png
+        COMMAND cp icon_x256.png  ${ICON_TARGET}/icon_256x256.png
+        COMMAND cp icon_x512.png  ${ICON_TARGET}/icon_256x256@2x.png
+        COMMAND cp icon_x512.png  ${ICON_TARGET}/icon_512x512.png
+        COMMAND cp icon_x1024.png ${ICON_TARGET}/icon_512x512@2x.png
+        COMMAND iconutil -c icns ${ICON_TARGET}
+        WORKING_DIRECTORY ${SOURCE_ICON_DIR})
+
+    set(MACOSX_BUNDLE_COPYRIGHT "OpenLoco is licensed under the MIT License")
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Maybe move the static stuff straight into the template?

Comment (AaronVanGeffen) [reply CMakeLists.txt]: Personally, I rather like that the template is reusable in its current form. Perhaps we should order these properties alphabetically, though.

Comment (marijnvdwerf) [reply CMakeLists.txt]: That's what Xcode seems to be doing as well: 
image

**Comment by marijnvdwerf** [comment]
Solves #223

File: CMakeLists.txt [hunk 9]
```diff
@@ -230,6 +231,53 @@ endif()
...
+        COMMAND iconutil -c icns ${ICON_TARGET}
+        WORKING_DIRECTORY ${SOURCE_ICON_DIR})
+
+    set(BUNDLE_RESOURCES ${ICON_OUTPUT})
+    list(APPEND BUNDLE_RESOURCES ${CMAKE_CURRENT_SOURCE_DIR}/CHANGELOG.md)
+    list(APPEND BUNDLE_RESOURCES ${CMAKE_CURRENT_SOURCE_DIR}/CONTRIBUTORS.MD)
+    set_target_properties(${PROJECT} PROPERTIES RESOURCE "${BUNDLE_RESOURCES}")
+
+    file(GLOB BUNDLE_LANGUAGES "${CMAKE_CURRENT_SOURCE_DIR}/data/language/*")
+    target_sources(${PROJECT} PUBLIC ${BUNDLE_LANGUAGES} ${BUNDLE_RESOURCES})
```
Comment (marijnvdwerf) [inline CMakeLists.txt]: Split this up in one call for the languages, and one for the other resources. Keeps everything a bit more split

File: src/openloco/platform/platform.h [hunk 10]
```diff
@@ -16,4 +16,7 @@ namespace openloco::platform
     std::string prompt_directory(const std::string& title);
     fs::path GetCurrentExecutablePath();
     std::vector<fs::path> get_drives();
+#if defined(__APPLE__) && defined(__MACH__)
+    fs::path GetBundlePath();
```
Comment (marijnvdwerf) [inline src/openloco/platform/platform.h]: Should probably start with a lowercase `g`

File: distribution/macos/Info.plist.in [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	<!--High Resolution Capable-->
+	<key>NSHighResolutionCapable</key>
+	<true/>
+
+	<!--Copyright (human-readable)-->
+	<key>NSHumanReadableCopyright</key>
+	<string>${MACOSX_BUNDLE_COPYRIGHT}</string>
+
+	<!--Localized resources can be mixed-->
+	<key>CFBundleAllowMixedLocalizations</key>
```
Comment (marijnvdwerf) [inline distribution/macos/Info.plist.in]: This one is at the wrong spot, but I think it can be removed entirely

**Comment by AaronVanGeffen** [comment]
Heads-up: the language files are added to the bundle, but the list is not populated properly.

**Comment by marijnvdwerf** [comment]
My local builds work fine. Downloaded builds indeed have the issue mentioned. I'm wondering if it tries to build with the homebrew-installed boost (those lines need to be removed, since we provide a static version now)

**Comment by AaronVanGeffen** [comment]
Yes, removing the homebrew boost line fixed the build for me, too.

**Pull request merged**